### PR TITLE
Add notification example using Telegram

### PR DESCRIPTION
**Pull request by lamw**
Closes: #583
Signed-off-by: William Lam <wlam@vmware.com>

## Summary

Example PowerCLI function to send Telegram message when a VM is vMotion from a specific vSphere Resource Pool

screenshots-01

## Pull Request Checklist
🚨 Please review the [guidelines for contributing](https://vmweventbroker.io/community) to this repository.

- [ ] Please ensure that you are making a pull request against the **Development** branch
- [ ] Please use the `WIP` keyword in the title of your PR if you are not ready for review
- [ ] Please ensure that you have opened a Github Issue if you are resolving/fixing a problem
- [ ] Please ensure that you have [signed](https://help.github.com/en/github/authenticating-to-github/signing-commits) all commits and that you have [squashed](https://medium.com/@slamflipstrom/a-beginners-guide-to-squashing-commits-with-git-rebase-8185cf6e62ec) all relevant commits related to your change
- [ ] Please make sure that you have tested your change locally by successfully [building and deploying the VMware Event Broker Appliance](https://vmweventbroker.io/kb/contribute-appliance) and/or [building and deploying VMware Event Router](https://vmweventbroker.io/kb/contribute-eventrouter)
- [ ] Please include any relevant screenshots and/or output as part of your testing
- [ ] Please include any documentation updates that is applicable for your changes

## Change Type

What types of changes does your code introduce to the VMware Event Broker Appliance?

_Put an `x` in all boxes that apply_

Please check the type of change your PR introduces:
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] Documentation changes
- [ ] Other (please describe):

### CODE REVIEW

File: examples/knative/powercli/kn-pcli-telegram/README.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+Change into the `test` directory
+```console
+cd test
+```
+
+Update the following variable names within the `docker-test-env-variable` file
+
+* TELEGRAM_BOT_API_KEY - Telegram Bot API key
+* TELEGRAM_GROUP_CHAT_ID - Telegram Group Chat ID
+* VCENTER_RESOURCE_POOL_FILTER - Name of vSphere Resource Pool to filter for and only sent Telegram if migrated VM is located in RP
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-telegram/README.md:29]:
```suggestion
* VCENTER_RESOURCE_POOL_FILTER - Name of vSphere Resource Pool to filter for and only send Telegram if migrated VM is located in RP
```

File: examples/knative/powercli/kn-pcli-telegram/README.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
+# kn-pcli-telegram
+Example Knative PowerCLI function for sending a [Telegram](https://telegram.org/) message when a VM is migrated and also resides in a specific vSphere Resource Pool.
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-telegram/README.md:2]: I understand the script will also work if VM is not part of RP right?

If so, suggesting to reword:

```suggestion
Example Knative PowerCLI function for sending a [Telegram](https://telegram.org/) message when a VM is migrated and optionally filter by its location in a vSphere Resource Pool.
```

File: examples/knative/powercli/kn-pcli-telegram/README.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+```
+
+In a separate terminal, run either `send-cloudevent-test.ps1` (PowerShell Script) or `send-cloudevent-test.sh` (Bash Script) to simulate a CloudEvent payload being sent to the local container image
+
+```console
+Testing Function ...
+See docker container console for output
+
+# Output from docker container console
+08/11/2021 14:45:44 - PowerShell HTTP server start listening on 'http://*:8080/'
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-telegram/README.md:48]: I guess these logs are from a <1.3 base image tag?

File: examples/knative/powercli/kn-pcli-telegram/function.yaml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+kind: Trigger
+metadata:
+  name: veba-pcli-telegram-trigger
+  labels:
+    app: veba-ui
+spec:
+  broker: default
+  filter:
+    attributes:
+      type: com.vmware.event.router/event
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-telegram/function.yaml:33]: Feel free to drop

File: examples/knative/powercli/kn-pcli-telegram/handler.ps1 [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+   $sendTelegram = $true
+
+   # Ignore ResourcePool filter if not provided
+   if(${jsonSecrets}.VCENTER_RESOURCE_POOL_FILTER -ne $NULL) {
+
+      # Construct VM Object given the MoRef ID from CloudEvent
+      $moRef = New-Object VMware.Vim.ManagedObjectReference
+      $moRef.Type = "VirtualMachine"
+      $moRef.Value = ${cloudEventData}.Vm.Vm.Value
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-telegram/handler.ps1:87]: I would move these lines inside the try/catch

File: examples/knative/powercli/kn-pcli-telegram/handler.ps1 [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+      if(${env:FUNCTION_DEBUG} -eq "true") {
+         Write-Host "$(Get-Date) - DEBUG: Telegram URL:`n`t${telegramUrl}`n"
+
+         Write-Host "$(Get-Date) - DEBUG: Telegram Message:`n`t${telegramMessage}`n"
+      }
+
+      try {
+         Invoke-WebRequest -Method POST -Uri $telegramUrl -ContentType "application/json;charset=utf-8" -Body (ConvertTo-Json -Compress -InputObject @{chat_id="$(${jsonSecrets}.TELEGRAM_GROUP_CHAT_ID)";text=$telegramMessage})
+      } catch {
+         throw "$(Get-Date) - Failed to send SMS: $($_.Exception.Response.StatusCode.value__)"
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-telegram/handler.ps1:128]: Why not throw whole exception, i.e `$_` ?

File: examples/knative/powercli/kn-pcli-telegram/telegram_secret.json [hunk 7]
```diff
@@ -0,0 +1,9 @@
+{
+  "TELEGRAM_BOT_API_KEY": "FILL-ME-IN",
+  "TELEGRAM_GROUP_CHAT_ID": "FILL-ME-IN",
+  "VCENTER_RESOURCE_POOL_FILTER": "FILL-ME-IN",
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-telegram/telegram_secret.json:4]:
```suggestion
  "VCENTER_RESOURCE_POOL_FILTER": "FILL-ME-IN-OR-REMOVE-THIS-LINE",
```

File: examples/knative/powercli/kn-pcli-telegram/telegram_secret.json [hunk 8]
```diff
@@ -0,0 +1,9 @@
...
+  "VCENTER_SERVER": "FILL-ME-IN",
+  "VCENTER_USERNAME" : "FILL-ME-IN",
+  "VCENTER_PASSWORD" : "FILL-ME-IN",
+  "VCENTER_CERTIFICATE_ACTION" : "Ignore"
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-telegram/telegram_secret.json:8]: Default should never be insecure

File: examples/knative/powercli/kn-pcli-telegram/test/docker-test-env-variable [hunk 9]
```diff
@@ -0,0 +1 @@
+TELEGRAM_SECRET={"TELEGRAM_BOT_API_KEY": "FILL-ME-IN","TELEGRAM_GROUP_CHAT_ID": "FILL-ME-IN","VCENTER_RESOURCE_POOL_FILTER": "FILL-ME-IN","VCENTER_SERVER": "FILL-ME-IN","VCENTER_USERNAME" : "FILL-ME-IN","VCENTER_PASSWORD" : "FILL-ME-IN","VCENTER_CERTIFICATE_ACTION" : "Ignore"}
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-telegram/test/docker-test-env-variable:1]: See suggested changes in comment of JSON file

**Review by embano1 — CHANGES_REQUESTED** [review]
Nits, but looks really good overall.

File: examples/knative/powercli/kn-pcli-telegram/README.md (hunk 3 above)
Comment (lamw) [reply examples/knative/powercli/kn-pcli-telegram/README.md:48]: Correct

File: examples/knative/powercli/kn-pcli-telegram/function.yaml (hunk 4 above)
Comment (lamw) [reply examples/knative/powercli/kn-pcli-telegram/function.yaml:33]: Sorry, which line am I dropping? I see L30-33 highlighted and I thought we needed 30-32 to be able to filter on a specific VC Event?

Comment (embano1) [reply examples/knative/powercli/kn-pcli-telegram/function.yaml:33]: 33 since 34 is specific enough IMHO

**Review by embano1 — APPROVED** [review]
LGTM

**Comment by lamw** [comment]
All feedback addressed

**Review by embano1 — APPROVED** [review]
LGTM

**Pull request merged**